perf-regression/gpu-margo-p2p-bw.cu

```diff
@@ -0,0 +1,796 @@
+        || (my_mpi_rank == 1 && g_opts.mercury_timeout_client == 0)) {
+        /* If mercury timeout of zero is requested, then set
+         * init option to NO_BLOCK.  This allows some transports to go
+         * faster because they do not have to set up or maintain the data
+         * structures necessary for signaling completion on blocked
+         * operations.
+         */
+        hii.na_init_info.progress_mode = NA_NO_BLOCK;
+    }
+
+    /* actually start margo */
+    mid = margo_init_ext(g_opts.na_transport, MARGO_SERVER_MODE, &mii);
+    assert(mid);
+
+    if (g_opts.diag_file_name) margo_diag_start(mid);
+
+    /* adjust mercury timeout in Margo if requested */
+    if (my_mpi_rank == 0 && g_opts.mercury_timeout_server != UINT_MAX) {
+        char timeout_s[64];
+        snprintf(timeout_s, 64, "%u", g_opts.mercury_timeout_server);
+        margo_set_param(mid, "progress_timeout_ub_msec", timeout_s);
+    }
+    if (my_mpi_rank == 1 && g_opts.mercury_timeout_client != UINT_MAX) {
+        char timeout_s[64];
+        snprintf(timeout_s, 64, "%u", g_opts.mercury_timeout_client);
+        margo_set_param(mid, "progress_timeout_ub_msec", timeout_s);
+    }
+
+    g_bw_id = MARGO_REGISTER(mid, "bw_rpc", bw_rpc_in_t, bw_rpc_out_t, bw_ult);
+
+    /* set up group */
+    ret = ssg_init();
+    assert(ret == SSG_SUCCESS);
+
+    if (my_mpi_rank == 0) {
+        /* set up server "group" on rank 0 */
+        ret = ssg_group_create_mpi(mid, "margo-p2p-bw", MPI_COMM_SELF, NULL,
+                                   NULL, NULL, &gid);
+        assert(ret == SSG_SUCCESS);
+
+        /* load group info into a buffer */
+        ssg_group_id_serialize(gid, 1, &gid_buffer, &gid_buffer_size);
+        assert(gid_buffer && (gid_buffer_size > 0));
+        gid_buffer_size_int = (int)gid_buffer_size;
+    }
+
+    /* broadcast server group info to clients */
+    MPI_Bcast(&gid_buffer_size_int, 1, MPI_INT, 0, MPI_COMM_WORLD);
+    if (my_mpi_rank == 1) {
+        /* client ranks allocate a buffer for receiving GID buffer */
+        gid_buffer = (char *)calloc((size_t)gid_buffer_size_int, 1);
+        assert(gid_buffer);
+    }
+    MPI_Bcast(gid_buffer, gid_buffer_size_int, MPI_CHAR, 0, MPI_COMM_WORLD);
+
+    if (my_mpi_rank == 1) {
+        int count = 1;
+        ssg_group_id_deserialize(gid_buffer, gid_buffer_size_int, &count, &gid);
+        assert(gid != SSG_GROUP_ID_INVALID);
+
+        ret = ssg_group_refresh(mid, gid);
+        assert(ret == SSG_SUCCESS);
+    }
+
+    /* sanity check group size on server/client */
+    ret = ssg_get_group_size(gid, &group_size);
+    assert(ret == SSG_SUCCESS);
+    assert(group_size == 1);
+
+    if (my_mpi_rank == 0) {
+        /* server side: prep everything before letting the client initiate
+         * benchmark
+         */
+        void* buffer = g_buffer;
+
+        /* register memory for xfer */
+
+	bulk_attr.mem_type = (hg_mem_type_t)NA_MEM_TYPE_CUDA;
```

Comment:
Should we only be setting the mem_type as NA_MEM_TYPE_CUDA in the case where we're actually using a GPU buffer?

Reply:
Fixed